Sources/Transaction/Memos/RecoverableMemo.swift

```diff
@@ -0,0 +1,74 @@
+//
+//  Copyright (c) 2020-2021 MobileCoin. All rights reserved.
+//
+
+import Foundation
+
+
+enum RecoveredMemo {
+    case sender(SenderMemo)
+    case destination(DestinationMemo)
+    case senderWithPaymentRequest(SenderWithPaymentRequestMemo)
+}
+
+enum RecoverableMemo {
+    case notset
+    case unused
+    case sender(RecoverableSenderMemo)
+    case destination(RecoverableDestinationMemo)
+    case senderWithPaymentRequest(RecoverableSenderWithPaymentRequestMemo)
+    
+    init(decryptedMemo data: Data66, accountKey: AccountKey, txOutKeys: TxOut.Keys) {
+        guard let memoData = Data64(data[2...]) else {
+            logger.warning("Memo data type unavailable")
+            self = .notset
+            return
+        }
+        let typeBytes = data[..<2]
+        
+        switch typeBytes.hexEncodedString() {
+        case Types.SENDER_WITH_PAYMENT_REQUEST:
+            let memo = RecoverableSenderWithPaymentRequestMemo(memoData, accountKey: accountKey, txOutPublicKey: txOutKeys.publicKey)
+            self = .senderWithPaymentRequest(memo)
+        case Types.SENDER:
+            let memo = RecoverableSenderMemo(memoData, accountKey: accountKey, txOutPublicKey: txOutKeys.publicKey)
+            self = .sender(memo)
+        case Types.DESTINATION:
+            let memo = RecoverableDestinationMemo(memoData, accountKey: accountKey, txOutKeys: txOutKeys)
+            self = .destination(memo)
```

Comment:
One thought here... 

Each of the "Revoverable<Memotype>()" functions can possibly return nil if the data is not valid...  

When that happens, we'll still be returning a non-nil RecoverableDemo enum for one of the valid types, but the memo itself will be nil...   

Almost wondering if we should add an 'invalid' case to use when the memo isn't parseable b/c of bad data?

Reply:
As a logical comparison, if the decryption fails, we set the value to .notSet in TxOutMemoParser...

Reply:
the constructors always return valid data. At one point they were nillable but I changed that by assuming all "addressHash" ffi functions return data, which is true, they generally always succeed with some sort of value because its just a computation on bytes.